libarchive/archive_read_support_format_tar.c

```diff
@@ -1323,28 +1338,60 @@ header_common(struct archive_read *a, struct tar *tar,
 		archive_entry_set_mtime(entry, tar_atol(header->mtime, sizeof(header->mtime)), 0);
 	}
 
-	/* Update size information as appropriate */
-	if (!archive_entry_size_is_set(entry)) {
-		tar->entry_bytes_remaining = tar_atol(header->size, sizeof(header->size));
-		if (tar->entry_bytes_remaining < 0) {
-			tar->entry_bytes_remaining = 0;
-			archive_set_error(&a->archive, ARCHIVE_ERRNO_MISC,
-					  "Tar entry has negative size");
-			return (ARCHIVE_FATAL);
-		}
-		if (tar->entry_bytes_remaining > entry_limit) {
-			tar->entry_bytes_remaining = 0;
-			archive_set_error(&a->archive, ARCHIVE_ERRNO_MISC,
-					  "Tar entry size overflow");
-			return (ARCHIVE_FATAL);
-		}
-		if (!tar->realsize_override) {
-			tar->realsize = tar->entry_bytes_remaining;
-		}
-		archive_entry_set_size(entry, tar->realsize);
-	} else if (tar->realsize_override) {
-		tar->entry_bytes_remaining = tar->realsize;
-		archive_entry_set_size(entry, tar->realsize);
+	/* Reconcile the size info. */
+	/* First, how big is the file on disk? */
+	if ((tar->size_fields & TAR_SIZE_GNU_SPARSE_REALSIZE) != 0) {
+		/* GNU sparse format 1.0 uses `GNU.sparse.realsize`
+		 * to hold the size of the file on disk. */
+		tar->disk_size = tar->GNU_sparse_realsize;
+	} else if ((tar->size_fields & TAR_SIZE_GNU_SPARSE_SIZE) != 0
+		   && (tar->sparse_gnu_major == 0)) {
```

Comment:
It's strange that `GNU.sparse.size` means completely different things in different versions of GNU tar, but at least they did put a version stamp so we can decide how to interpret it.